File: OneMap/DuckDB/scripts/import_excel_duckdb.py
#!/usr/bin/env python3
"""
DuckDB Excel Import Script
Uses pandas to read Excel and DuckDB for fast analytics
"""

import sys
import os
import pandas as pd
import duckdb
from pathlib import Path
from datetime import datetime

# Configuration
DB_PATH = Path(__file__).parent.parent / "data" / "onemap.duckdb"
DB_PATH.parent.mkdir(exist_ok=True)

def import_excel_to_duckdb(excel_path, sheet_name='Sheet1'):
    """Import Excel file to DuckDB using pandas"""
    
    print(f"🦆 DuckDB Excel Import (Python)")
    print(f"📄 File: {excel_path}")
    print(f"📋 Sheet: {sheet_name}")
    
    # Check if file exists
    if not os.path.exists(excel_path):
        print(f"❌ File not found: {excel_path}")
        return False
    
    # Get file size
    file_size_mb = os.path.getsize(excel_path) / (1024 * 1024)
    print(f"📊 File size: {file_size_mb:.2f} MB")
    
    try:
        # Read Excel file with pandas
        print("\n📥 Reading Excel file...")
        df = pd.read_excel(excel_path, sheet_name=sheet_name)
        print(f"✅ Loaded {len(df)} rows, {len(df.columns)} columns")
        
        # Clean column names (remove special characters)
        df.columns = [col.replace(' ', '_').replace(':', '').replace('-', '_') for col in df.columns]
        
        # Show column info
        print("\n📋 Columns found:")
        for i, col in enumerate(df.columns):
            print(f"   {i+1}. {col} ({df[col].dtype})")
        
        # Create DuckDB connection
        print(f"\n🦆 Connecting to DuckDB: {DB_PATH}")
        con = duckdb.connect(str(DB_PATH))
        
        # Create table from DataFrame
        table_name = "excel_import"
        print(f"\n📥 Importing to table: {table_name}")
        con.execute(f"DROP TABLE IF EXISTS {table_name}")
        con.execute(f"CREATE TABLE {table_name} AS SELECT * FROM df")
        
        # Get import statistics
        row_count = con.execute(f"SELECT COUNT(*) FROM {table_name}").fetchone()[0]
        print(f"✅ Successfully imported {row_count} rows")
        
        # Show sample data
        print("\n📄 Sample data (first 5 rows):")
        sample = con.execute(f"SELECT * FROM {table_name} LIMIT 5").df()
        print(sample)
        
        # Analyze the data
        analyze_data(con, table_name, df.columns)
        
        # Create useful views
        create_views(con, df.columns)
        
        # Close connection
        con.close()
        
        print("\n✅ Import complete!")
        print("\n💡 To query the data:")
        print(f"   duckdb {DB_PATH}")
        print(f"   SELECT * FROM {table_name} LIMIT 10;")
        
        return True
        
    except Exception as e:
        print(f"\n❌ Error: {e}")
        return False

def analyze_data(con, table_name, columns):
    """Analyze the imported data"""
    print("\n📊 Data Analysis:")
    
    # Look for status columns
    status_cols = [col for col in columns if 'status' in col.lower()]
    if status_cols:
        for col in status_cols:
            print(f"\n📈 Distribution of '{col}':")
            result = con.execute(f"""
                SELECT "{col}", COUNT(*) as count
                FROM {table_name}
                WHERE "{col}" IS NOT NULL
                GROUP BY "{col}"
                ORDER BY count DESC
                LIMIT 10
            """).df()
            print(result)
    
    # Look for date columns
    date_cols = [col for col in columns if 'date' in col.lower()]
    if date_cols:
        for col in date_cols:
            try:
                result = con.execute(f"""
                    SELECT 
                        MIN("{col}") as earliest,
                        MAX("{col}") as latest
                    FROM {table_name}
                    WHERE "{col}" IS NOT NULL
                """).fetchone()
                print(f"\n📅 Date range for '{col}':")
                print(f"   Earliest: {result[0]}")
                print(f"   Latest: {result[1]}")
            except:
                pass
    
    # Look for agent/pole columns
    agent_cols = [col for col in columns if 'agent' in col.lower()]
    pole_cols = [col for col in columns if 'pole' in col.lower()]
    
    if agent_cols:
        col = agent_cols[0]
        print(f"\n👥 Top 10 by '{col}':")
        result = con.execute(f"""
            SELECT "{col}", COUNT(*) as count
            FROM {table_name}
            WHERE "{col}" IS NOT NULL
            GROUP BY "{col}"
            ORDER BY count DESC
            LIMIT 10
        """).df()
        print(result)
    
    if pole_cols:
        col = pole_cols[0]
        unique_count = con.execute(f"""
            SELECT COUNT(DISTINCT "{col}") 
            FROM {table_name}
            WHERE "{col}" IS NOT NULL
        """).fetchone()[0]
        print(f"\n📍 Unique {col}: {unique_count}")

def create_views(con, columns):
    """Create useful views for analysis"""
    print("\n📊 Creating analysis views...")
    
    # Create a summary view if we have status and date columns
    status_col = next((col for col in columns if 'status' in col.lower()), None)
    date_col = next((col for col in columns if 'date' in col.lower()), None)
    
    if status_col and date_col:
        con.execute(f"""
            CREATE OR REPLACE VIEW status_summary AS
            SELECT 
                DATE_TRUNC('month', "{date_col}") as month,
                "{status_col}" as status,
                COUNT(*) as count
            FROM excel_import
            WHERE "{date_col}" IS NOT NULL
            GROUP BY month, status
            ORDER BY month DESC, count DESC
        """)
        print("✅ Created view: status_summary")
    
    print("✅ Views created successfully")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python import_excel_duckdb.py <excel_file> [sheet_name]")
        print("Example: python import_excel_duckdb.py data/onemap.xlsx 'Sheet1'")
        sys.exit(1)
    
    excel_file = sys.argv[1]
    sheet_name = sys.argv[2] if len(sys.argv) > 2 else 'Sheet1'
    
    success = import_excel_to_duckdb(excel_file, sheet_name)
    sys.exit(0 if success else 1)

File: OneMap/DuckDB/scripts/import-excel-duckdb.js
#!/usr/bin/env node

const Database = require('duckdb-async').Database;
const path = require('path');
const fs = require('fs');

// Configuration
const DB_PATH = path.join(__dirname, '../data/onemap.duckdb');

// Command line arguments
const args = process.argv.slice(2);
if (args.length === 0) {
    console.log('Usage: node import-excel-duckdb.js <excel-file-path> [sheet-name]');
    console.log('Example: node import-excel-duckdb.js ../data/1754473447790_Lawley_01082025.xlsx');
    process.exit(1);
}

const excelPath = path.resolve(args[0]);
const sheetName = args[1] || 'Sheet1';

// Validate file exists
if (!fs.existsSync(excelPath)) {
    console.error('❌ Excel file not found:', excelPath);
    process.exit(1);
}

console.log('🦆 DuckDB Excel Import Tool (Async)');
console.log('📄 File:', excelPath);
console.log('📋 Sheet:', sheetName);
console.log('📊 File size:', (fs.statSync(excelPath).size / 1024 / 1024).toFixed(2), 'MB');

async function importExcel() {
    const db = await Database.create(DB_PATH);
    
    try {
        // Install and load spatial extension (includes Excel support in newer versions)
        console.log('\n📦 Installing extensions...');
        await db.run('INSTALL spatial');
        await db.run('LOAD spatial');
        
        // Create a simple import table from the Excel file
        console.log('\n📥 Creating import table from Excel...');
        
        // First, let's try a different approach - use CSV reading capability
        // Many Excel files can be read as CSV-like structures
        const createTableQuery = `
            CREATE OR REPLACE TABLE excel_import AS 
            SELECT * FROM read_csv_auto('${excelPath.replace(/'/g, "''")}')
        `;
        
        try {
            await db.run(createTableQuery);
            console.log('✅ Data imported using CSV auto-detection');
        } catch (csvError) {
            console.log('⚠️  CSV auto-detection failed, trying Parquet approach...');
            
            // Alternative: Convert Excel to temporary CSV using a different method
            console.error('❌ Direct Excel import not available in this DuckDB version');
            console.log('\n💡 Alternative approaches:');
            console.log('1. Convert Excel to CSV first');
            console.log('2. Use Python with pandas + DuckDB');
            console.log('3. Use the SQLite system in OneMap/SQL/');
            throw csvError;
        }
        
        // Get row count
        const countResult = await db.all('SELECT COUNT(*) as count FROM excel_import');
        console.log(`\n✅ Successfully imported ${countResult[0].count} rows`);
        
        // Show table structure
        console.log('\n📋 Table Structure:');
        const columns = await db.all('DESCRIBE excel_import');
        console.table(columns);
        
        // Show sample data
        console.log('\n📄 Sample Data (first 5 rows):');
        const sample = await db.all('SELECT * FROM excel_import LIMIT 5');
        console.table(sample);
        
        // Analyze the data
        await analyzeData(db);
        
    } catch (error) {
        console.error('\n❌ Import error:', error.message);
        
        // Suggest using the SQLite system instead
        console.log('\n💡 Recommendation: Use the SQLite-based system instead');
        console.log('   cd OneMap/SQL/scripts');
        console.log('   npm run import ../data/excel/1754473447790_Lawley_01082025.xlsx');
        
    } finally {
        await db.close();
    }
}

async function analyzeData(db) {
    console.log('\n📊 Data Analysis:');
    
    try {
        // Get column names
        const columns = await db.all("SELECT column_name FROM information_schema.columns WHERE table_name = 'excel_import'");
        const columnNames = columns.map(c => c.column_name);
        
        // Look for status columns
        const statusCol = columnNames.find(col => col.toLowerCase().includes('status'));
        if (statusCol) {
            const statusCounts = await db.all(`
                SELECT "${statusCol}" as status, COUNT(*) as count
                FROM excel_import
                GROUP BY "${statusCol}"
                ORDER BY count DESC
                LIMIT 10
            `);
            console.log(`\n📈 Status Distribution (${statusCol}):`);
            console.table(statusCounts);
        }
        
        // Look for date columns
        const dateCol = columnNames.find(col => col.toLowerCase().includes('date'));
        if (dateCol) {
            const dateRange = await db.all(`
                SELECT 
                    MIN("${dateCol}") as earliest,
                    MAX("${dateCol}") as latest,
                    COUNT(DISTINCT "${dateCol}") as unique_dates
                FROM excel_import
            `);
            console.log(`\n📅 Date Range (${dateCol}):`);
            console.table(dateRange[0]);
        }
        
        // Show all available columns
        console.log('\n📋 Available Columns for Analysis:');
        columnNames.forEach((col, idx) => {
            console.log(`   ${idx + 1}. ${col}`);
        });
        
        console.log('\n✅ Import and analysis complete!');
        console.log('\n💡 Next steps:');
        console.log('1. Query the data: SELECT * FROM excel_import WHERE ...');
        console.log('2. Export results: COPY (SELECT ...) TO "output.csv" (HEADER, DELIMITER ",")');
        console.log('3. Create analytics views for common queries');
        
    } catch (error) {
        console.error('⚠️  Analysis error:', error.message);
    }
}

// Run the import
importExcel().catch(console.error);